### PR TITLE
Migrate from CircleCI to GitHub Actions for testing and releases, enable linting and changelog

### LINKED ISSUE #794: Enable go-changelog Automation
Similar to other Go modules maintained by the SDK team, this project's CHANGELOG handling should be switched to [`go-changelog`](https://github.com/hashicorp/go-changelog), including:

- Adding `.changelog` directory and template files
- Enabling automation for regenerating the CHANGELOG (e.g. scripts or GitHub Actions)
- (If enhancements are made available upstream in time) Enabling automation for checking CHANGELOG entries and formatting

### PR DESCRIPTION
**Pull request by bflad**
Closes #794

All testing and linting workflows verified as successful via:

```shell
act -s GITHUB_TOKEN=$GITHUB_TOKEN pull_request
```

### CODE REVIEW

File: .github/CONTRIBUTING.md [hunk 1]
```diff
@@ -100,11 +102,159 @@ In the case of `terraform-plugin-sdk`, the repo's close relationship to the `ter
 
 We belive that one should "leave the campsite cleaner than you found it", so you are welcome to clean up cosmetic issues in the neighbourhood when submitting a patch that makes functional changes or fixes.
 
+### Dependency Updates
+
+Dependency management is performed by [dependabot](https://docs.github.com/en/code-security/supply-chain-security/keeping-your-dependencies-updated-automatically/about-dependabot-version-updates). Where possible, dependency updates should occur through that system to ensure all Go module files are appropriately updated and to prevent duplicated effort of concurrent update submissions. Once available, updates are expected to be verified and merged to prevent latent technical debt.
+
+### Changelog
+
+HashiCorp’s open-source projects have always maintained user-friendly, readable CHANGELOGs that allow practitioners and developers to tell at a glance whether a release should have any effect on them, and to gauge the risk of an upgrade. We use the [go-changelog](https://github.com/hashicorp/go-changelog) to generate and update the changelog from files created in the `.changelog/` directory.
```
Comment (bookshelfdave) [inline .github/CONTRIBUTING.md:111]: `We use the` -> `We use`

**Review by bookshelfdave — APPROVED** [review]
LGTM. Lots of content and all I could find were a few minor typos 😸

**Review by CalebAlbers — APPROVED** [review]
The GitHub Actions workflows look good to me! It looks like there is a workflow to validate goreleaser config, but not to run goreleaser for a given tag/release - is that ran manually right now?

File: .github/CONTRIBUTING.md [hunk 2]
```diff
@@ -100,11 +102,159 @@ In the case of `terraform-plugin-sdk`, the repo's close relationship to the `ter
...
+
+``````markdown
+```release-note:feature
+Added `great` package, which solves all the problems
+```
+``````
+
+###### Bug Fixes
+
+A new bug entry should use the `release-note:bug` header and have a prefix indicating the sub-package it corresponds to, a colon, then followed by a brief summary. Use a `all` prefix should the fix apply to all sub-packages.
```
[the remark this answers is not in the document]
Comment (bflad) [reply .github/CONTRIBUTING.md]: Fixed! Thanks for catching that.

File: .github/CONTRIBUTING.md (hunk 1 above)
Comment (bflad) [reply .github/CONTRIBUTING.md]: Fixed! Thanks for catching that.

**Comment by bflad** [comment]
@CalebAlbers good catch! Copied the release workflow over as well.

**Pull request merged**